java/ql/src/Security/CWE/CWE-020/OverlyLargeRange.qhelp

```diff
@@ -0,0 +1,74 @@
+<!DOCTYPE qhelp PUBLIC
+"-//Semmle//qhelp//EN"
+"qhelp.dtd">
+<qhelp>
+
+    <overview>
+        <p>
+            A regexp range can by accident match more than was intended.
+            For example, the regular expression <code>/[a-zA-z]/</code> will 
+            match every lowercase and uppercase letters, but the same regular 
+            expression will also match the chars: <code>[\]^_`</code>.  
+        </p>
+        <p>
+            On other occasions it can happen that the dash in a regular 
+            expression is not escaped, which will cause it to be interpreted 
+            as part of a range. For example in the character class <code>[a-zA-Z0-9%=.,-_]</code> 
+            the last character range matches the 55 characters between 
+            <code>,</code> and <code>_</code> (both included), which overlaps with the 
+            range <code>[0-9]</code> and is thus clearly not intended.
+        </p>
+    </overview>
+
+    <recommendation>
+        <p>
+
+            Don't write character ranges were there might be confusion as to
+            which characters are included in the range.
+
+        </p>
+    </recommendation>
+
+    <example>
+
+        <p>
+            The following example code checks whether a string is a valid 6 digit hex color. 
+        </p>
+
+<sample language="java">
+import java.util.regex.Pattern
+public class Tester {
+    public static boolean is_valid_hex_color(String color) {
+        return Pattern.matches("#[0-9a-fA-f]{6}", color);
+    }
+}
+</sample>
+
+        <p>
+            However, the <code>A-f</code> range matches every uppercase character, and
+            thus a "color" like <code>#XYZ</code> is considered valid.
```

Comment:
Suggest:

```suggestion
            However, the <code>A-f</code> range is overly large and matches every uppercase character. It would parse 
            a "color" like <code>#XYZ</code> as valid.
```

Reply:
However, `#XYZ` is not 6 digits, so possibly this needs to be:

```suggestion
            However, the <code>A-f</code> range is overly large and matches every uppercase character. It would parse 
            a "color" like <code>#XXYYZZ</code> as valid.
```